src/Microsoft.ML.Tokenizers/Utils/DoubleArrayTrie.cs

```diff
@@ -0,0 +1,1138 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// The implementation of the DoubleArrayBuilder class is based on the following C# port of the C++ implementation of the Double-Array Trie (DART) data structure.
```

Comment:
Should we unit test this type?  Lots of complexity here.  Might be easier to maintain with direct tests.

Reply:
I'm not worried about it not being tested. I suggested this because it's an individual unit that might be easier to hit all behaviors with unit tests rather than functional tests.   Thank your for adding the tests.